model-lowerings-common/src/org/jetbrains/dukat/model/commonLowerings/overrides/VarConflictResolveLowering.kt

```diff
@@ -0,0 +1,223 @@
+            if (with (typeChecker) {
+                    val parentType = (parentProperty.memberModel as PropertyModel).type
+                    member.type.isOverridingReturnType(parentType) && !member.type.isEquivalent(parentType)
+                }) {
+                return parentProperty
+            }
+        }
+        return null
+    }
+
+    private fun findSuperType(types: List<TypeModel>, typeChecker: OverrideTypeChecker): TypeModel? {
+        var superType: TypeModel? = types[0]
+        types.forEach { type ->
+            if (with(typeChecker) { superType?.isOverridingReturnType(type) } == true) {
+                superType = type
+            } else if (superType != null && with(typeChecker) { !type.isOverridingReturnType(superType!!) }) {
+                superType = null
+            }
+        }
+        return superType
+    }
+
+    private fun createPropertyToAdd(properties: List<MemberData>, classLike: ClassLikeModel): PropertyModel? {
+        val name = (properties[0].memberModel as PropertyModel).name
+        return if (classLike.members.filterIsInstance<PropertyModel>().any { it.name == name }) {
+            null
+        } else {
+            val superType = findSuperType(
+                properties.map { (it.memberModel as PropertyModel).type },
+                OverrideTypeChecker(modelContext, inheritanceContext, classLike, null)
+            )
+            superType?.let { type ->
+                (properties[0].memberModel as PropertyModel).copy(
+                    type = type,
+                    override = properties[0].fqName
+                ).toVal()
+            }
+        }
+    }
+
+    private fun generateNewMembers(
+        allConflictingProperties: List<List<MemberData>>,
+        varPropertiesWithSpecifiedType: List<MemberData>,
+        classLike: ClassLikeModel
+    ): List<MemberModel> {
+
+        val propertiesToAdd = allConflictingProperties.mapNotNull { createPropertyToAdd(it, classLike) }
+
+        return classLike.members.mapNotNull { member ->
+            if (member is PropertyModel) {
+                when {
+                    allConflictingProperties.any { properties ->
+                        (properties[0].memberModel as PropertyModel).name == member.name
+                    } -> {
+                        member.toVal()
+                    }
+                    varPropertiesWithSpecifiedType.any { (it.memberModel as PropertyModel).name == member.name } -> {
+                        member.toVal()
+                    }
+                    else -> {
+                        member
+                    }
+                }
+            } else {
+                member
+            }
+        } + propertiesToAdd
+    }
+
+    override fun lowerClassLikeModel(
+        ownerContext: NodeOwner<ClassLikeModel>,
+        parentModule: ModuleModel
+    ): ClassLikeModel? {
+        val classLike = super.lowerClassLikeModel(ownerContext, parentModule) ?: return null
```

Comment:
I'm against of using return in this particular case, let's discuss. First thing - can we make lowerClassLikeModel non-nullable?

Reply:
we actually can, fixed